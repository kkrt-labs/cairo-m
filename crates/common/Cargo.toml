[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
