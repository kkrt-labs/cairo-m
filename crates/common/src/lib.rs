pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
