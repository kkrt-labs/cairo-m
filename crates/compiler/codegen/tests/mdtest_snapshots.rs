//! Codegen snapshot tests for mdtest cases.
//! This file automatically generates codegen snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the entire compilation pipeline.

mod common;

use cairo_m_compiler_codegen::CodeGenerator;
use cairo_m_compiler_mir::generate_mir;
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_codegen_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("CASM", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            // First generate MIR
            let mir_module = match generate_mir(&db, crate_id) {
                Ok(module) => module,
                Err(diagnostics) => {
                    return Err(format!("MIR generation failed: {:#?}", diagnostics));
                }
            };

            // Then generate CASM code from MIR
            let mut generator = CodeGenerator::new();
            match generator.generate_module(&mir_module) {
                Ok(_) => Ok(generator.debug_instructions()),
                Err(e) => Err(format!("Code generation failed: {:#?}", e)),
            }
        })
        .with_parent_dir(true);

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("Codegen snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
