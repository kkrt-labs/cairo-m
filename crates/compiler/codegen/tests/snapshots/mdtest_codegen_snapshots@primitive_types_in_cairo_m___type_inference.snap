---
source: crates/compiler/codegen/tests/mdtest_snapshots.rs
description: "Codegen snapshot for mdtest: Primitive Types in Cairo-M - Type Inference"
input_file: mdtest/01-basics/03-primitive-types.md
---
Source:
fn test_inference() -> u32 {
    let x = 42;  // Type inferred as felt
    let y: u32 = 100;
    let z = 200u32;
    return y + z;
}
============================================================
Generated CASM:
test_inference:
test_inference:
test_inference_0:
   0: 23 100 0 0           // u32([fp + 0, fp + 1]) = u32(100)
   1: 23 200 0 2           // u32([fp + 2, fp + 3]) = u32(200)
   2: 15 0 2 4             // u32([fp + 4], [fp + 5]) = u32([fp + 0], [fp + 1]) op u32([fp + 2], [fp + 3])
   3: 4 4 0 -4             // Return value 0 slot 0: [fp -4] = [fp + 4] + 0
   4: 4 5 0 -3             // Return value 0 slot 1: [fp -3] = [fp + 5] + 0
   5: 11 _ _ _             // return
